standalone/prouter/src/i2pd.rs

```diff
@@ -60,77 +63,93 @@ pub fn generate_ident_to_file(
     let c_str_filename = CString::new(String::from(keyfile_path.to_string_lossy()))
         .expect("String should be able to be converted into CString");
     let c_filename: *const c_char = c_str_filename.as_ptr() as *const c_char;
-    let ptr_ident: *const c_char = unsafe {
+    let mut buf = [0 as c_char; BUFFER_SIZE];
+    let ret = unsafe {
         C_GenerateIdentToFile(
             c_filename,
             c_sk,
             PRouterSigningKeyType::SigningKeyTypeEDDSASha512Ed25519 as u16,
             PRouterCryptoKeyType::CryptoKeyTypeElgamal as u16,
+            buf.as_mut_ptr(),
+            buf.len() as u64,
         )
     };
 
-    if ptr_ident.is_null() {
+    if ret != 0 {
         error!("The ident returned from C is corrupted");
         return Err(anyhow!("The ident returned from C is corrupted"));
     }
 
-    let c_str_ident = unsafe { CStr::from_ptr(ptr_ident) };
-    // Ok(c_str_ident.to_bytes().to_owned())
-    let ident: String = c_str_ident.to_str()?.to_owned();
-
-    Ok(ident)
+    unsafe { String::from_utf8(std::mem::transmute(buf.to_vec())) }.map_err(|e| {
+        anyhow!(
+            "Failed to convert the value returned from C to String: {:?}",
```

Comment:
1. We should still use `CStr::from_ptr` to read the result buf because the content is indeed a c string. `String::from_utf8` would take the padding `\0`s into account.
2. Better make helper function to do the ffi str to String convert.

Moreover, the C side changes have 2 issues:

https://github.com/Soptq/i2pd/compare/67d0b17162ad8412de950a1274c40d24a4b285ee...205698f507a945d5078519f335787cc7b5746686#diff-5c9126c541227786b59e490b2dd7a7a67521f266710722bf7efa64d3ceb5b351R102-R109

1. it still strncpy strings into the global `RET_STR` rather than the `buf`.
2. strncpy is not safe. If the buf_len is less than the actual source string len, it would not set the tail `\0`. Using `snprintf` can be safer.

Reply:
Fixed ;)